CLI: uploading videos, creating channels, updating videos and channels

Lezek123 opened the pull request:
Implements: https://github.com/Joystream/joystream/issues/1443
Based on: https://github.com/Joystream/joystream/pull/1509

Commands added:
- `media:createChannel`
- `media:updateChannel`
- `media:uploadVideo PATH`
- `media:updateVideo`
- `media:myVideos`
- `media:myChannels`

Commands modified:
- `content-directory:class` - added `Index` (`PropertyID`) column
- `content-directory:entities` - added `--filters` flag allowing for basic filtering of the results. 

_See updated `README` at: https://github.com/Lezek123/substrate-runtime-joystream/blob/cli-media/cli/README.md for further description of the new commands. Note that most of the arguments/flags are optional, in which case entities can be selected via a CLI choice-list prompt_

#### Uploading a video - features
- `ffmpeg.ffprobe` is used to retrieve some information (metadata) about the video, like: codec name, width, height and duration, and provide them as defaults / hints when prompting for `Video` / `VideoMedia` entity values. https://www.npmjs.com/package/@ffmpeg-installer/ffmpeg is used here for cross-platform support. Since it seems like this approach may be prone to some errors and is not essential for the command to do its job, the failure to retrieve this data will only result in a warning (the command itself will not be terminated).
- To resolve the upload url - either the local ipfs node will be used (if available) or random bootstrap discovery endpoint (fetched via `api.query.discovery.bootstrapEndpoints`). The information about which one was used will be logged in the console.
- There is some basic error handling for cases like too large file or no storage provider available. Other errors (like invalid mime type) are handled during the upload (caught from `axios.put` when sending a request to upload endpoint)
- Specifying custom user license is not yet implemented (currently only one of the `KnownLicense`s can be chosen)

Commit:
@joystream/types - poroposalDetails and WorkingGroup enum update (Baylon)

Commit:
generate:all

Commit:
media: commands family - channels and videos

Commit:
Storage node - just enough changes to make it work

Commit:
yarn.lock & Readme update

Commit:
Merge branch 'content-directory-schemas' into cli-media

Lezek123 committed:
Merge branch 'babylon' into cli-media

Commit:
Add back ProposalDetails enum variants

Commit:
"media:myVideos", "media:myChannels" + few improvements

Commit:
uploadVideo - filePath as arg

Commit:
Readme update

Commit:
Merge branch 'babylon' into cli-media

Commit:
Merge branch 'babylon' into cli-media

Commit:
https://github.com/Joystream/joystream/pull/1509#issuecomment-709213195 fixes

Commit:
Save output before sending extrinsic and normalize ExtrinsicFailed to be errors

mnaamani commented on cli/src/base/AccountsCommandBase.ts:
```diff
@@ -224,6 +224,16 @@ export default abstract class AccountsCommandBase extends ApiCommandBase {
+      this.error('Membership required to access this command!', { exit: ExitCodes.AccessDenied })
+    }
+
+    return memberIds[0].toNumber() // FIXME: Temporary solution (just using the first one)
```
I wanted to raise this issue in previous PR about member id selection and clearly you are aware of it hence your "FIXME" comment. Just wanted to assert that yes it would be good to give use option to select the member id of their choice when there are multiple matches..

mnaamani commented on cli/src/base/ApiCommandBase.ts:
```diff
@@ -405,11 +406,26 @@ export default abstract class ApiCommandBase extends StateAwareCommandBase {
+                  try {
+                    // Need to assert that registry is of TypeRegistry type, since Registry intefrace
+                    // seems outdated and doesn't include DispatchErrorModule as possible argument for "findMetaError"
+                    const { name, documentation } = (this.getOriginalApi().registry as TypeRegistry).findMetaError(
```
wonderful, I will try to update storage-node with same code, there i just pass the raw dispatch error which is quite useless :)

mnaamani commented:
Wroked great. Uploaded a video successfully and played it back over asset url.
And nice fixes from last PR too.

Just couple observations:

- When prompted to enter video metadata after upload:
```
File uploaded!
 Choose Video encoding MPEG4
? pixelWidth 
? pixelHeight 
 Choose Video language English
 Choose Video category Cartoon
 Choose License CC_BY
? title my first video
? description zoom 4
? thumbnailURL 
? duration dunno.  
? isPublic Yes
? isExplicit No
? hasMarketing No
```

for `duration` I entered text without being warned that it should be a number.

- Inspecting `Video` entity shows a strange values for `curation status`, `skippableIntroDuration`, `publishedBeforeJoystream`
```
MokoBook:joystream mokhtar$ yarn joystream-cli content-directory:entity 29
yarn run v1.22.4
$ /Users/mokhtar/joystream/joystream/node_modules/.bin/joystream-cli content-directory:entity 29
ID                            29        
Class name                    Video     
Class ID                      11        
Supported schemas             [0]       
Controller                    Member(0) 
Frozen                        false     
Refrecencable                 true      
Same owner references         0         
Total references              0         

________________ Property values _________________

channel                       24 Single<Reference>        
category                      2 Single<Reference>         
title                         my first video Single<Text> 
description                   zoom 4 Single<Text>         
duration                      0 Single<Uint32>            
skippableIntroDuration        false Single<Uint16>   // expected number
thumbnailURL                   Single<Text>               
language                      10 Single<Reference>        
media                         28 Single<Reference>        
hasMarketing                  false Single<Bool>          
publishedBeforeJoystream      false Single<Uint32>       // uhm ?
isPublic                      true Single<Bool>           
isExplicit                    false Single<Bool>          
license                       25 Single<Reference>        
curationStatus                false Single<Reference>.   // expected entity id but showing 
```

- Prompt for account password
I'm always prompted for password for account which doesn't have one. perhaps function that loads the key can try to unlock the kerying without a password first or check if it is locked, only if it fails to unlock then prompt for a password? Less hassle for user.

I'll try to do a bit more testing like with no ipfs node, storage node being down...

Commit:
Review fixes - https://github.com/Joystream/joystream/pull/1516#issuecomment-709446079

Lezek123 commented:
Added a few fixes:
- > for duration I entered text without being warned that it should be a number.
   
  Fixed in `JSONSchemaPrompt` by constructing `Ajv` with `allErrors: true` (which is required when we prompt for values in non- 
  standard order,  ie. by using `promptMultipleProps`) and changing normalizers for `integer` and `number` to only return the normalized value if it matches the original input when converted `.toString` 
- > Inspecting Video entity shows a strange values for curation status, skippableIntroDuration, publishedBeforeJoystream
  
  When the value for a property is not set (which can be the case when the property is not `required`), the runtime seems to return `Bool(false)`. I added an additional check to handle this case in commands like `entity` and `entities` and display a more clear message (`[not set]`)
- > I'm always prompted for password for account which doesn't have one

Fixed + added a loop when prompting for password (when it is actually set to non-empty string), so that providing invalid one doesn't exit the command right away (which I found to also be quite bad UX)

- > I'll try to do a bit more testing like with no ipfs node, storage node being down...

I noticed a small error in `isAnyProviderAvailable` check which I probably introduced accidently at some point, but it should be fixed now.

mnaamani commented on cli/src/Api.ts:
```diff
@@ -481,16 +490,18 @@ export default class Api {
+  async availableClasses(useCache = true): Promise<[ClassId, Class][]> {
+    return useCache && this._cdClassesCache
+      ? this._cdClassesCache
+      : await this.entriesByIds<ClassId, Class>(this._api.query.contentDirectory.classById)
```
shouldn't we set `_cdClassesCache` here and then return it after querying it from chain. I don't see any place where it is ever set.

Lezek123 replied:
Right, I overlooked that, thank you for noticing!

mnaamani commented on cli/src/base/AccountsCommandBase.ts:
```diff
@@ -216,12 +216,34 @@ export default abstract class AccountsCommandBase extends ApiCommandBase {
 
   async requestAccountDecoding(account: NamedKeyringPair): Promise<void> {
-    const password: string = await this.promptForPassword()
+    // First - try decoding using empty string
```
We could also return early if `!account.isLocked`

Lezek123 replied:
Good idea, added the check.

mnaamani commented on cli/src/base/ContentDirectoryCommandBase.ts:
```diff
@@ -163,4 +244,33 @@ export default abstract class ContentDirectoryCommandBase extends AccountsComman
     }, {} as Record<string, { value: Codec; type: string }>)
   }
+
+  async parseToKnownEntityJson<T>(entity: Entity, entityClass?: Class): Promise<FlattenRelations<T>> {
```
It seems safer to never take the optional argument `entityClass` since it can be fetched by class_id from Entity.
`parseEntityPropertyValues()` will not give proper result if wrong entityClass is passed to it, and that could be hard to debug?

Lezek123 replied:
That's true, I actually forgot to remove this, the purpose of the optional argument was to avoid an additional api call, but after I implmeneted `_cdClassesCache` it is no longer necessary.

mnaamani commented on cli/src/helpers/InputOutput.ts:
```diff
@@ -4,9 +4,13 @@ import ExitCodes from '../ExitCodes'
 import chalk from 'chalk'
 
+// Default schema path for resolving refs
+const DEFAULT_SCHEMA_PATH = getSchemasLocation('entities') + path.sep
```
Here we are assuming that the schemas we have at this location match the schemas on chain.
That may not always a valid expectation and the cli should probably not rely on them, but instead read them/reconstruct them from the chain.

mnaamani replied:
Also on this point, if cli package depends on the cd-schemas package, we would have to make sure to publish that to npm as well. Is it something we want to do?

Lezek123 replied:
I'm not sure if this still applies after https://github.com/Joystream/joystream/pull/1530, but:

1. I think it would be very useful to have `cd-schemas` published either way. Perhaps it can become part of `@joystream/js` if I manage to get into that before the release. If not, it can be just a standalone library that makes interactions with `content-directory` runtime module easier.

1. > Here we are assuming that the schemas we have at this location match the schemas on chain.

Since this is used by higher-level commands like `uploadVideo` etc. that make some assumptions about the `Video` schema anyway (when populating `VideoEntity` interface with data provided by the user) I think it may be acceptable? 

The alternative would be to avoid any assumptions about classes/schemas at all in the CLI and just have a command like `createEntity Video 0` (where `0` would be `schemaIndex`), This was the idea behind https://github.com/Joystream/joystream/issues/1441 which I postponed a bit due to change of priorities, but I hope we'll have both of those solutions implmented soon and we could then compare their pros and cons (for now I think having a nice, customizable UI for uploading a video is a good thing, because CLI will temporarly be the only place where this could be done)

mnaamani commented:
> Added a few fixes:
> 
>     * > for duration I entered text without being warned that it should be a number.
>       
>       
>       Fixed in `JSONSchemaPrompt` by constructing `Ajv` with `allErrors: true` (which is required when we prompt for values in non-
>       standard order,  ie. by using `promptMultipleProps`) and changing normalizers for `integer` and `number` to only return the normalized value if it matches the original input when converted `.toString`
> 
>     * > Inspecting Video entity shows a strange values for curation status, skippableIntroDuration, publishedBeforeJoystream
>       
>       
>       When the value for a property is not set (which can be the case when the property is not `required`), the runtime seems to return `Bool(false)`. I added an additional check to handle this case in commands like `entity` and `entities` and display a more clear message (`[not set]`)
> 
>     * > I'm always prompted for password for account which doesn't have one
> 
> 
> Fixed + added a loop when prompting for password (when it is actually set to non-empty string), so that providing invalid one doesn't exit the command right away (which I found to also be quite bad UX)
> 
>     * > I'll try to do a bit more testing like with no ipfs node, storage node being down...
> 
> 
> I noticed a small error in `isAnyProviderAvailable` check which I probably introduced accidently at some point, but it should be fixed now.

Tested :+1:

Commit:
Review fixes

Commit:
Merge branch 'babylon' into cli-media

Commit:
Yarn.lock fix

Commit:
Remove curator froum group

Lezek123 commented:
Added missing `removeCuratorFromGroup` command

Commit:
Metadata cache

Lezek123 commented:
Added a small api initialization speed improvement by implementing metadata cache, I was expecting commands to execute ~2-3 seconds faster after that, but turned out it only saves ~300-800 ms on local chain (which is still something).

I suspect the difference may be greater when fetching data from a remote node.

mnaamani commented:
> Added a small api initialization speed improvement by implementing metadata cache, I was expecting commands to execute ~2-3 seconds faster after that, but turned out it only saves ~300-800 ms on local chain (which is still something).
> 
> I suspect the difference may be greater when fetching data from a remote node.

Nice addition. Yes might be more noticeable when using a remote endpoint.
